docs/rfcs/0005-ecommerce-rest-api.md

```diff
@@ -0,0 +1,208 @@
+time a user PATCHes this endpoint, the basket will be validated. If the basket is not valid
+a 400 error will be returned and the basket will be left unchanged.
+
+Also note that a basket passing validation is not a guarantee that the checkout will
+pass validation. This could happen if the coupon code was valid when it was being
+added to the basket but it was not valid by the time of checkout. A product may also
+have its validity changed, preventing checkout from working.
+
+Whatever the request data sent to PATCH, if the response is a success it will contain
+the same updated data which GET would return.
+
+TODO: what happens if a basket becomes invalid over time? I think this is somewhat rare
+but we should figure out how to handle it. We can just clear the basket but I don't know when
+we should do this validation.
+
+##### Clear a basket
+
+A user could clear a basket by patching:
+
+    {
+        "items": [],
+        "coupons": []
+    }
+
+They could clear just the items in the basket:
+
+    {
+        "items": []
+    }
+    
+However this would leave the coupon code as is. If there was a coupon code it would
+fail validation because there is nothing to apply the coupon to, and the basket would
+be left unchanged. So there is no use case for this.
+
+A user could clear only the coupon:
+
+    {
+        "coupons": []
+    }
+
+However, if there is an automatic coupon which can be applied, it will be automatically
+added to the basket even when `coupons` is explicitly set to an empty list.
+    
+
+##### Add item to basket
+
+A user could update the basket to have one item by patching:
+
+    {
+        "items": [
+            {
+                "id": 8   // this is ProductVersion.id, not Product.id
+            }
+        ]
+    }
+    
+Here the coupon is not one of the keys so it would be left as is. However the view
+will revalidate the coupon to make sure it can be applied to the new item list. If this
+validation fails a 400 response is returned and the basket is left unchanged.
+
+Also note:
+ - There is no way to incrementally add items to a basket, leaving existing
+items unchanged. The frontend will need to send the complete new list with `id` attributes
+for each product. However in practice we will not have more than one item in a basket.
+ - At the moment a user can purchase only one item at a time. This will be validated
+ on the backend.
+
+
+##### Add coupon to basket
+
+A user could apply a coupon to a basket:
+
+    {
+        "coupons": [
+            {
+                "code": "DISCOUNT_30_OFF"
+            }
+        ]
+    } 
+
+Existing items in the basket would be left unchanged. The new coupon code would be validated.
+If the validation fails a 400 response would be returned and the basket would be left
+unchanged.
+
+### /api/checkout/
+
+Unlike MicroMasters there is no payload required for the checkout API. Instead it will
+use the user's basket to figure out what should be purchased.
```

Comment:
I want to be sure that we can still pass order data to the merchant-defined fields in cybersource.

Reply:
From the docs it looks like we can have up to 100 merchant-defined fields so the number of items in the basket will need to be limited depending on how many we use per basket item.

Reply:
Actually it looks like the first four are stored differently from the other 96, though I'm not sure what the significance is for us:

>Merchant-defined data fields 1 to 4 are stored against the payment token and are used for subsequent token based transactions. Merchant defined data fields 5 to 100 are passed trough to Decision Manager as part of the initial payment request and are not stored against the payment token. 
From page 96 of [the docs](http://apps.cybersource.com/library/documentation/dev_guides/Secure_Acceptance_Hosted_Checkout/Secure_Acceptance_Hosted_Checkout.pdf)